### PR TITLE
[MOB-2899] - Health Monitoring for Offline Processor

### PR DESCRIPTION
**Pull request by Ayyanchira**
## 🔹 Jira Ticket(s) if any

https://iterable.atlassian.net/browse/MOB-2899

## ✏️ Description

1. Dont schedule for offline processing if row count exceeds max allowed number
2. Max offline operation is set to 1000
3. Added tests for Healthmonitoring and offline processing
4. DB error checks and fallback logic

### CODE REVIEW

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableTaskStorage.java [hunk 1]
```diff
@@ -254,13 +255,24 @@ private IterableTask createTaskFromCursor(Cursor cursor) {
         return taskIds;
     }
 
+    long numberOfTasks() throws Exception {
+        if (!isDatabaseReady()) {
+            notifyDBNotReady();
+            throw new Exception("Database is not ready");
```
Comment (davidtruong) [inline iterableapi/src/main/java/com/iterable/iterableapi/IterableTaskStorage.java:261]: Consider using a IllegalStateException to make the exception more specific.

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableTaskStorage.java [hunk 2]
```diff
@@ -445,14 +457,87 @@ private boolean updateTaskWithContentValues(String id, ContentValues contentValu
 
     private boolean isDatabaseReady() {
         if (database == null) {
+            notifyDBNotReady();
```
Comment (davidtruong) [inline iterableapi/src/main/java/com/iterable/iterableapi/IterableTaskStorage.java:460]: What's the use case for calling both notifyDBNotReady and notifyDBError?

Comment (roninopf) [reply iterableapi/src/main/java/com/iterable/iterableapi/IterableTaskStorage.java:460]: That's definitely an error, from before we formalized the error notifications. Removed.

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableConstants.java [hunk 3]
```diff
@@ -235,6 +235,9 @@
     // Custom actions handled by the SDK
     public static final String ITERABLE_IN_APP_ACTION_DELETE    = "delete";
 
+    //Offline operation
+    public static final long MAX_OFFLINE_OPERATION              = 1000;
```
Comment (vbabenkoru) [inline iterableapi/src/main/java/com/iterable/iterableapi/IterableConstants.java:239]: The name looks confusing.. What is 'max operation'?

File: iterableapi/src/main/java/com/iterable/iterableapi/HealthMonitor.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
+package com.iterable.iterableapi;
+
+public class HealthMonitor implements IterableTaskStorage.IterableDatabaseStatusListeners {
+    private static final String TAG = "HealthMonitor";
+
+    private boolean errored = false;
```
Comment (vbabenkoru) [inline iterableapi/src/main/java/com/iterable/iterableapi/HealthMonitor.java:6]: Naming: It's not clear what exactly errored and what this variable is for.

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableTaskStorage.java [hunk 5]
```diff
@@ -254,13 +255,24 @@ private IterableTask createTaskFromCursor(Cursor cursor) {
         return taskIds;
     }
 
+    long numberOfTasks() throws Exception {
```
Comment (vbabenkoru) [inline iterableapi/src/main/java/com/iterable/iterableapi/IterableTaskStorage.java:258]: Methods in Java are usually verbs: `getTasksCount`/`getNumberOfTasks`

File: iterableapi/src/main/java/com/iterable/iterableapi/HealthMonitor.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    public void isClosed() {
+        IterableLogger.v(TAG, "DB Closed notified to healthMonitor");
+        errored = true;
+    }
+
+    @Override
+    public void isReady() {
+        IterableLogger.v(TAG, "DB Ready notified to healthMonitor");
+        errored = false;
+    }
```
Comment (vbabenkoru) [inline iterableapi/src/main/java/com/iterable/iterableapi/HealthMonitor.java:67]: Do these have to be separate methods? Looks like they're doing the same thing. If the logic is the same, I think it's better to have one method, and if we need to have different messages in the log, a method parameter should be enough.

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableTaskStorage.java [hunk 7]
```diff
@@ -445,14 +457,86 @@ private boolean updateTaskWithContentValues(String id, ContentValues contentValu
 
     private boolean isDatabaseReady() {
```
Comment (vbabenkoru) [inline iterableapi/src/main/java/com/iterable/iterableapi/IterableTaskStorage.java:458]: Is `isDatabaseReady` the only place where we can detect DB errors? One of the failing tests was crashing in `createTask` when calling `database.insert`, and that wasn't being caught by the health monitor. I think we should catch errors in every method that calls the database and log that and flip the flag. The current implementation probably won't catch database corruption errors or low storage errors.

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableTaskStorage.java [hunk 8]
```diff
@@ -445,14 +457,86 @@ private boolean updateTaskWithContentValues(String id, ContentValues contentValu
...
+        void onDBError();
+
+        void isClosed();
+
+        void isReady();
+    }
+
+    private ArrayList<IterableDatabaseStatusListeners> databaseStatusListeners = new ArrayList<>();
+
+    void addDataBaseListener(IterableDatabaseStatusListeners listener) {
```
Comment (vbabenkoru) [inline iterableapi/src/main/java/com/iterable/iterableapi/IterableTaskStorage.java:486]: * `database` is one word, no need to capitalize as `DataBase`.
* "Database listener" sounds too generic, I think "database state listener" or "database error listener" would be more specific.

File: iterableapi/src/main/java/com/iterable/iterableapi/OfflineRequestProcessor.java [hunk 9]
```diff
@@ -70,7 +78,7 @@ boolean isRequestOfflineCompatible(String baseUrl) {
     }
 }
 
-//Placeholder Taskschedular for testing purpose.
+//Mock TaskScheduler for testing purpose.
```
Comment (vbabenkoru) [inline iterableapi/src/main/java/com/iterable/iterableapi/OfflineRequestProcessor.java:81]: Why is this a mock? Looks like a code that's used in production.

File: iterableapi/src/main/java/com/iterable/iterableapi/HealthMonitor.java (hunk 4 above)
Comment (Ayyanchira) [reply iterableapi/src/main/java/com/iterable/iterableapi/HealthMonitor.java:6]: Makes sense.. will name it databaseErrored

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableConstants.java (hunk 3 above)
Comment (Ayyanchira) [reply iterableapi/src/main/java/com/iterable/iterableapi/IterableConstants.java:239]: Maximum number of offline operations supported?
hmm.. may be OFFLINE_OPERATION_THRESHOLD or
OFFLINE_LIMIT or
OFFLINE_OPERATIONS_LIMIT

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableTaskStorage.java (hunk 2 above)
Comment (Ayyanchira) [reply iterableapi/src/main/java/com/iterable/iterableapi/IterableTaskStorage.java]: Removed similar methods and now it only notifies for two states, Ready and Error

File: iterableapi/src/main/java/com/iterable/iterableapi/HealthMonitor.java (hunk 6 above)
Comment (Ayyanchira) [reply iterableapi/src/main/java/com/iterable/iterableapi/HealthMonitor.java]: Changed it to notify only two states -> errored and ready

File: iterableapi/src/main/java/com/iterable/iterableapi/OfflineRequestProcessor.java (hunk 9 above)
Comment (Ayyanchira) [reply iterableapi/src/main/java/com/iterable/iterableapi/OfflineRequestProcessor.java]: Removed the line

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableTaskStorage.java (hunk 7 above)
Comment (Ayyanchira) [reply iterableapi/src/main/java/com/iterable/iterableapi/IterableTaskStorage.java:461]: Added code to read the returned value when inserting into database and then requesting via online mode.
Also flagging Healthmonitor in such cases

**Review by davidtruong — APPROVED** [review]
LFTM

**Review by vbabenkoru — APPROVED** [review]

**Pull request merged**